### PR TITLE
Avoid `Read`/`WriteBigEndian`

### PR DESCRIPTION
**Pull request by cdesouza-chromium**
These functions are getting deleted by upstream and getting replaced by safer version from `//base/numerics`. This change corrects all occurrences of these functions.

Chromium change:
https://chromium.googlesource.com/chromium/src/+/437342e0e2e8759f07c7de53e7512d202d761a83

commit 437342e0e2e8759f07c7de53e7512d202d761a83
Author: danakj <danakj@chromium.org>
Date:   Tue Mar 12 14:35:27 2024 +0000

    Remove base::{Read,Write}BigEndian

    These are now dead code, and replaced by base::numerics byte
    conversions.

    R=tsepez@chromium.org

    Bug: 40284755

Resolves

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by zenparsing — APPROVED** [review]

File: components/brave_wallet/common/zcash_utils.cc [hunk 1]
```diff
@@ -48,19 +49,13 @@ std::optional<uint64_t> ReadCompactSize(base::span<const uint8_t>& data) {
     value = type;
     data = data.subspan(1);
   } else if (type == 253 && data.size() >= 3) {
-    uint16_t val = 0;
-    base::ReadBigEndian(&data[1], &val);
-    value = val;
+    value = base::numerics::U16FromBigEndian(data.subspan(1).first<2u>());
     data = data.subspan(1 + 2);
   } else if (type <= 254 && data.size() >= 5) {
```
Comment (supermassive) [inline components/brave_wallet/common/zcash_utils.cc:54]: is `subspan + first` some pattern from chrome?
why not just like `subspan(1, 2u)` everywhere?

Comment (supermassive) [reply components/brave_wallet/common/zcash_utils.cc:54]: or subspan<1, 2u>() if we need compile time sized span

**Review by supermassive — APPROVED** [review]
wallet core lgtm with some notes

File: components/brave_wallet/common/zcash_utils.cc (hunk 1 above)
Comment (cdesouza-chromium) [reply components/brave_wallet/common/zcash_utils.cc]: Done.

File: components/brave_wallet/common/btc_like_serializer_stream.cc [hunk 2]
```diff
@@ -15,21 +16,18 @@ void BtcLikeSerializerStream::Push8AsLE(uint8_t i) {
 }
 
 void BtcLikeSerializerStream::Push16AsLE(uint16_t i) {
-  i = base::ByteSwapToLE16(i);
-  base::span<uint8_t> data_to_insert(reinterpret_cast<uint8_t*>(&i), sizeof(i));
-  PushBytes(data_to_insert);
+  PushBytes(base::byte_span_from_ref(
+      base::numerics::U16FromLittleEndian(base::byte_span_from_ref(i))));
```
Comment (supermassive) [inline components/brave_wallet/common/btc_like_serializer_stream.cc:20]: ByteSwapToLE16 convert *to* LE and is a noop for LE platforms(btw do we have one supported?)
Now we always covert *from* LE which seem to work but semantically incorrect

Comment (supermassive) [reply components/brave_wallet/common/btc_like_serializer_stream.cc:20]: TIL we actually always on LE platforms https://chromium-review.googlesource.com/c/chromium/src/+/5369340

Let's then that be `PushBytes(base::byte_span_from_ref(i))`

File: components/brave_wallet/browser/zcash/zcash_serializer.cc [hunk 3]
```diff
@@ -29,10 +31,15 @@ constexpr uint32_t kConsensusBranchId = 0xC2D6D0B4;
...
+  std::string personalizer(kTxHashPersonalizerPrefix);
+
+  personalizer.append('\0', sizeof(kConsensusBranchId));
+  base::as_writable_byte_span(personalizer)
+      .subspan<std::size(kTxHashPersonalizerPrefix),
+               sizeof(kConsensusBranchId)>()
+      .copy_from(base::byte_span_from_ref(base::numerics::U32FromLittleEndian(
+          base::byte_span_from_ref(kConsensusBranchId))));
+
   return personalizer;
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_serializer.cc:43]: Let's replace all of this with
```
// https://zips.z.cash/zip-0244#txid-digest-1
// https://github.com/zcash/librustzcash/blob/3620314a/zcash_primitives/src/transaction/txid.rs#L377-L380
// https://github.com/zcash/librustzcash/blob/3620314a/components/zcash_protocol/src/consensus.rs#L573
constexpr char kTxHashPersonalizer[] = "ZcashTxHash_\xB4\xD0\xD6\xC2";
```

**Pull request closed without merging**